content/purgatorio/launch-and-mantain-the-system/automation-and-reproducibility.md

```diff
@@ -1,5 +1,242 @@
+
+    This issue often arises when using Jupyter Notebooks to build preprocessing workflows, watch out! Rather consider solid alternatives such as the [SKLearn's Pipelines](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html).
+
+    Be sure to check these out (coming from [Ten Simple Rules collection](https://collections.plos.org/ten-simple-rules)):
+
+    - [**Ten Simple Rules for Reproducible Research**](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003285)
+    - [**Ten simple rules for writing and sharing computational analyses in Jupyter Notebooks**](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1007007)
+
+    Interesting reads on Reproducibility:
+
+    - [Creating Reproducible Data Science Projects](https://towardsdatascience.com/creating-reproducible-data-science-projects-1fa446369386)
+    - [Reproducibility, Replicability, and Data Science](https://www.kdnuggets.com/2019/11/reproducibility-replicability-data-science.html)
+    - [Challenges in irreproducible research](https://www.nature.com/collections/prbfkwmwvz)
+
+- **Model training and serving**
+
+    You can create scripts to _automatically_ train ML models at any given time, perhaps using new data you have collected, cleaned, and pre-processed. 
+
+    You can automatically check that the performance of the models reflects expectations, and train them again in case they degrade.
+    
+    See:
+    - [Why Machine Learning Models Degrade In Production](https://towardsdatascience.com/why-machine-learning-models-degrade-in-production-d0f2108e9214)
+    - [The Ultimate Guide to Model Retraining](https://mlinproduction.com/model-retraining/)
+
+    Once you've retrained the model, you can _automatically_ deploy it in production. Many Cloud Providers (AWS, Azure, GCP) offer you the possibility to do this in a very simple way.
+    
+#### The Automation Golden Rule
+
+As you can see, there's a big room for improvement in your Data Science work if you use automation.
+
+The [Rule of Three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) in software development says: 
+
+> It states that two instances of similar code don't require refactoring, but when a similar code is used three times, it should be extracted into a new procedure.
+
+_Likewise_, you can think of automation!
+
+**If you should take a step more than twice, it's probably worth taking a moment to automate it!**
+
+## AutoML
+
+Until now we have seen how automation can fit into the various steps of the Data Science process, speed them up, and how it is vital to obtain reproducible results. 
+
+But there is another aspect of how automation can help us, and it is called [AutoML](https://en.wikipedia.org/wiki/Automated_machine_learning). 
+
+Automated machine learning (AutoML) is the process of automating the process of applying machine learning to real-world problems.
```

Comment:
Btw you want to know that AutoML may create a lot of work at times, that needs checking and verification. 

Many would still prefer to check their models manually or through some rigorous testing in case automation is chosen as an end-to-end process. Maybe creating these checks and balances can be part of the process, and if a new model isn't as performant as the previous one, there is a rollback performed and the concerned stakeholders are informed of this.

Reply:
Totally agree, one should find a good balance respect to the business case